File: clojure-euler/src/clojure_euler/problem4.clj
;A palindromic number reads the same both ways. The
;largest palindrome made from the product of two 2-digit
;numbers is 9009 = 91 99.

;Find the largest palindrome made from the product of two
;3-digit numbers.

;==> 906609
(ns clojure-euler.problem4
  (:require [clojure.string :as string]))

(defn palindrome? [x]
  (= (str x) (clojure.string/join (reverse (str x)))))

(defn solve-problem4 []
  (apply max (filter palindrome? (for [x (range 100 1000)
                                     y (range 100 1000)]
                                 (* x y)))))

File: ruby-euler/problem4.rb
# A palindromic number reads the same both ways. The
# largest palindrome made from the product of two 2-digit
# numbers is 9009 = 91 99.
#
# Find the largest palindrome made from the product of two
# 3-digit numbers.
#
# ==> 906609

require './toolkit'
include Toolkit

def palindrome?(val)
  val.to_s == val.to_s.reverse
end

max_palindrome = 0
min = 100
max = 999
res = 0

range_generator(min-1, max, true).each do |i|
  range_generator(min-1, i-1, true).each do |j|
    res = i * j
    max_palindrome = res if palindrome?(res) && res > max_palindrome
  end
end

puts max_palindrome
